mirrord/cli/src/error.rs

```diff
@@ -462,3 +480,144 @@ impl From<OperatorApiError> for CliError {
+    };
+
+    use http_body_util::Full;
+    use hyper::{
+        body::{Bytes, Incoming},
+        service::service_fn,
+        Request, Response,
+    };
+    use hyper_util::{
+        rt::{TokioExecutor, TokioIo},
+        server::conn::auto::Builder,
+    };
+    use k8s_openapi::api::core::v1::Pod;
+    use kube::{api::ListParams, Api};
+    use rustls::{crypto::aws_lc_rs::default_provider, pki_types::CertificateDer, ServerConfig};
+    use tokio::{net::TcpListener, sync::Notify};
+    use tokio_rustls::TlsAcceptor;
+
+    /// With this test we're trying to `assert` that our [`kube`] crate is (somewhat)
+    /// version-synced with [`rustls`]. To give a friendlier error message on kube requests
+    /// when there's a certificate problem, we must dig down into the [`kube::Error`].
+    ///
+    /// Certificate errors come under the [`kube::Error::Service`] variant, which holds
+    /// a very annoying type-erased boxed error. Trying to `downcast` this is messy, so
+    /// instead we check the debug message of the error.
+    ///
+    /// We want this error (or something similar) to happen:
+    ///
+    /// `Service(hyper_util::client::legacy::Error(Connect, Custom { kind: Other, error: Custom {
+    /// kind: InvalidData, error: InvalidCertificate(Expired) } }))`.
+    ///
+    /// The part that we care about is the `InvalidCertificate`, the rest is not relevant.
+    ///
+    /// This test may fail if [`rustls`] changes the `error: InvalidCertificate` message,
+    /// or any of the upper errors change.
+    ///
+    /// Relying on the `Debug` string version of the error, as the `Display` version is
+    /// just a generic `ServiceError: client error (Connect)`.
+    #[tokio::test]
+    async fn kube_service_error_dependency_is_in_sync() {
+        use kube::{Client, Config};
+
+        let provider = default_provider();
+        provider.install_default().unwrap();
+
+        let notify = Arc::new(Notify::new());
+        let wait_notify = notify.clone();
+
+        tokio::spawn(async move {
+            run_hyper_tls_server(notify).await;
+        });
+
+        // Wait until the server is listening before we start connecting.
+        wait_notify.notified().await;
+
+        let kube_config = Config::new("https://127.0.0.1:9669".parse().unwrap());
+        let client = Client::try_from(kube_config).unwrap();
+
+        // Manage pods
+        let pods: Api<Pod> = Api::default_namespaced(client);
+
+        let lp = ListParams::default().fields(&format!("legend={}", "hank"));
+
+        if let Err(fail) = pods.list(&lp).await {
+            assert!(format!("{fail:?}").contains("InvalidCertificate"),
+                "We were expecting this error {fail:?} to have `InvalidCertificate` but it's {fail}!"
+            );
+        }
+    }
+
+    /// Creates an [`hyper`] server with a broken certificate and a _catch-all_ route on
+    /// `localhost:9669`.
+    ///
+    /// The `.pem` files live in `/mirrord/tests/issue-2824-certs`.
```

Comment:
Can you maybe generate this cert dynamically? :eyes:

Reply:
Sure, but why? I grabbed this cert from the hyper examples repo.

Reply:
Dunno, just a suggestion. I don't like random files in the repo. But I don't insist ^^

Reply:
Changed!